docs/cs-proof.md

```diff
@@ -609,6 +609,115 @@ to prove that
 
 
+Padding \\(\mathbf{l}(x)\\) and \\(\mathbf{r}(x)\\) for the inner product proof
+-------------------------------------------------------------------------------
+
+The above discussion did not have restrictions on the value \\(n\\).
+However, the [inner product argument](../notes/index.html#inner-product-proof)
+requires the input vectors to have power-of-two elements: \\(n^{+} = 2^k\\).
+To resolve this mismatch, we need to specify how to pad vectors \\(\mathbf{l}(x), \mathbf{r}(x)\\)
+and related commitments before we can use the inner product argument.
+
+Our goal is to translate the _padding of the constraint system_ into the _padding of proof data_,
+so we can keep the constraint system small and perform less computations in proving and verification.
+
+We start by padding the entire constraint system:
+multipliers are padded with all-zero assignments \\(a\_{L,j}, a\_{R,j}, a\_{O,j}\\),
+all-zero blinding factors \\(s\_{L,j}, s\_{R,j}\\),
+and all-zero weights \\(W\_{R,i,j}, W\_{L,i,j}, W\_{O,i,j}\\),
+for all constraints \\(i \in [0, q)\\) and all additional multipliers \\(j \in [n,n')\\):
+
+\\[
+\begin{aligned}
+\mathbf{a}\_L^{+} &= \mathbf{a}\_L \hspace{0.1cm} || \hspace{0.1cm} [0,...,0] \\\\
+\mathbf{a}\_R^{+} &= \mathbf{a}\_R \hspace{0.1cm} || \hspace{0.1cm} [0,...,0] \\\\
+\mathbf{a}\_O^{+} &= \mathbf{a}\_O \hspace{0.1cm} || \hspace{0.1cm} [0,...,0] \\\\
+\mathbf{s}\_L^{+} &= \mathbf{s}\_L \hspace{0.1cm} || \hspace{0.1cm} [0,...,0] \\\\
+\mathbf{s}\_R^{+} &= \mathbf{s}\_R \hspace{0.1cm} || \hspace{0.1cm} [0,...,0] \\\\
+\mathbf{W}\_L^{+} &= \mathbf{W}\_L \hspace{0.1cm} || \hspace{0.1cm} [[0,...,0],...,[0,...,0]] \\\\
+\mathbf{W}\_R^{+} &= \mathbf{W}\_R \hspace{0.1cm} || \hspace{0.1cm} [[0,...,0],...,[0,...,0]] \\\\
+\mathbf{W}\_O^{+} &= \mathbf{W}\_O \hspace{0.1cm} || \hspace{0.1cm} [[0,...,0],...,[0,...,0]] \\\\
+\end{aligned}
+\\]
+
+As a result, the vectors of generators \\(\mathbf{G},\mathbf{H}\\) and challenges \\(\mathbf{y}^n\\) are extended:
+
+\\[
+\begin{aligned}
+\mathbf{G}^{+}     &= \mathbf{G}   \hspace{0.1cm} || \hspace{0.1cm} [G_n,...,G_{n^{+}-1}] \\\\
+\mathbf{H}^{+}     &= \mathbf{H}   \hspace{0.1cm} || \hspace{0.1cm} [H_n,...,H_{n^{+}-1}] \\\\
+\mathbf{y}^{n^{+}} &= \mathbf{y}^n \hspace{0.1cm} || \hspace{0.1cm} [y^n,...,y^{n^{+}-1}] \\\\
+\end{aligned}
+\\]
+
+The low-level variables are padded with zeroes, so their commitments remain unchanged:
+
+\\[
+\begin{aligned}
+A_I^{+} &= \widetilde{B} \cdot \tilde{a} + \langle \mathbf{G}^{+} , \mathbf{a}\_L^{+} \rangle + \langle \mathbf{H}^{+}, \mathbf{a}\_R^{+} \rangle &{}={}& A_I \\\\
+A_O^{+} &= \widetilde{B} \cdot \tilde{o} + \langle \mathbf{G}^{+} , \mathbf{a}\_O^{+} \rangle                                                     &{}={}& A_O \\\\
+S^{+}   &= \widetilde{B} \cdot \tilde{s} + \langle \mathbf{G}^{+} , \mathbf{s}\_L^{+} \rangle + \langle \mathbf{H}^{+}, \mathbf{s}\_R^{+} \rangle &{}={}& S
+\end{aligned}
```

Comment:
Might be slightly more clean if this shows the step of breaking the a_L+ into two parts, one of which is all zeros

Reply:
Done in 72fae27